cpp/src/arrow/util/io_util.cc

```diff
@@ -193,34 +210,79 @@ Status StringToNative(const std::string& s, NativePathString* out) {
   return Status::OK();
 }
 
+#if _WIN32
+Status NativeToString(const NativePathString& ws, std::string* out) {
+#if _WIN32
+  std::string s;
+  RETURN_NOT_OK(::arrow::util::WideStringToUTF8(ws, &s));
+  *out = std::move(s);
+#else
+  *out = ws;
+#endif
+  return Status::OK();
+}
+#endif
+
+#if _WIN32
+const wchar_t kNativeSep = L'\\';
+const wchar_t kGenericSep = L'/';
+const wchar_t* kAllSeps = L"\\/";
+#else
+const char kNativeSep = '/';
+const char kGenericSep = '/';
+const char* kAllSeps = "/";
+#endif
+
+NativePathString NativeSlashes(NativePathString s) {
+  return ReplaceChars(std::move(s), kGenericSep, kNativeSep);
+}
+
+NativePathString GenericSlashes(NativePathString s) {
+  return ReplaceChars(std::move(s), kNativeSep, kGenericSep);
+}
+
+NativePathString NativeParent(const NativePathString& s) {
+  // XXX What about trailing separators?
+  auto last_sep = s.find_last_of(kAllSeps);
+  if (last_sep == NativePathString::npos) {
+    return s;
+  }
+  // There may be multiple contiguous separators, skip all of them
+  auto before_seps = s.find_last_not_of(kAllSeps, last_sep);
+  if (before_seps == NativePathString::npos) {
+    // All separators are at start of string, keep them all
+    return s.substr(0, last_sep + 1);
+  } else {
+    return s.substr(0, before_seps + 1);
+  }
+}
+
+Status ValidatePath(const std::string& s) {
+  if (s.find_first_of('\0') != std::string::npos) {
+    return Status::Invalid("Embedded NUL char in path: '", s, "'");
+  }
+  return Status::OK();
+}
+
 }  // namespace
 
+#ifdef ARROW_WITH_BOOST_FILESYSTEM
+
+// NOTE: catching filesystem_error gives more context than system::error_code
+// (it includes the file path(s) in the error message)
+
 #define BOOST_FILESYSTEM_TRY try {
 #define BOOST_FILESYSTEM_CATCH           \
   }                                      \
   catch (bfs::filesystem_error & _err) { \
     return ToStatus(_err);               \
   }
 
-// NOTE: catching filesystem_error gives more context than system::error_code
-// (it includes the file path(s) in the error message)
-
 static Status ToStatus(const bfs::filesystem_error& err) {
   return Status::IOError(err.what());
 }
 
-static std::string MakeRandomName(int num_chars) {
-  static const std::string chars = "0123456789abcdefghijklmnopqrstuvwxyz";
-  std::random_device gen;
-  std::uniform_int_distribution<int> dist(0, static_cast<int>(chars.length() - 1));
-
-  std::string s;
-  s.reserve(num_chars);
-  for (int i = 0; i < num_chars; ++i) {
-    s += chars[dist(gen)];
-  }
-  return s;
-}
+#endif
```

Comment:
Would you mind decorating all multiline `#ifdef`/`#endif` blocks with comments?
```suggestion
#endif // ARROW_WITH_BOOST_FILESYSTEM
```

Reply:
Will do.